sky/serve/load_balancer.py

```diff
@@ -223,10 +322,250 @@ async def startup():
+        # received on the specified port using the filters described above.
+        # received on port 8080 and processes it with the above filters.
+        listener = {
+            'name': 'listener',
+            'address': {
+                'socket_address': {
+                    'address': '0.0.0.0',
+                    'port_value': self._load_balancer_port,
+                }
+            },
+            'filter_chains': [{
+                'filters': filters
+            }]
+        }
+
+        # These endpoints are the backend IP addresses that the cluster defined
+        # below can load balance over.
+        lb_endpoints = []
+        for url in replicas:
+            # TODO(ejj) when Envoy is the only load balancer choice, we should
+            # just have the controller send us the IP and port rather than
+            # bundling them together in a url.
+            parsed_url = urlparse(url)
+            lb_endpoints.append({
+                'endpoint': {
+                    'address': {
+                        'socket_address': {
+                            'address': parsed_url.hostname,
+                            'port_value': parsed_url.port,
+                        }
+                    }
+                }
+            })
+
+        # A cluster is usually a group of endpoints that can be load balanced
+        # over.  This cluster contains the replicas of the service, and is
+        # handed traffic by the above listener.
+        cluster = {
+            'name': 'cluster',
+            'connect_timeout': '0.25s',
+            'type': 'STATIC',
+            'lb_policy': 'ROUND_ROBIN',  # TODO(ejj) weighted least request
+            'load_assignment': {
+                'cluster_name': 'cluster',
+                'endpoints': [{
+                    'lb_endpoints': lb_endpoints
+                }]
+            },
+        }
+
+        config = {
+            'static_resources': {
+                'listeners': [listener],
+                'clusters': [cluster],
+            }
+        }
+
+        return yaml.dump(config, default_flow_style=False)
+
+    async def run(self):
+        await self.update_envoy_config([])
+
+        # Because docker can take some time to come up, we make multiple
+        # attempts to start before giving up.
+        envoy_started = False
+        logger.info('Starting envoy %s', self.envoy_name)
+        for _ in range(30):
+            await asyncio.sleep(5)
+            envoy_started = await self._start_envoy()
+            if envoy_started:
+                break
+
+        if not envoy_started:
+            error = f'Failed to start envoy {self.envoy_name}'
+            logger.error(error)
+            raise RuntimeError(error)
+
+        while True:
+            await asyncio.sleep(constants.LB_CONTROLLER_SYNC_INTERVAL_SECONDS)
+
+            ready_replica_urls = await self._controller_sync({})
```

Comment:
Just want to make sure, so currently we does not support report qps to the controller right? maybe worth a TODO here

Reply:
Ah yes I was planning to either implement this, or add a TODO before the formal PR

Reply:
SG!